File: app/src/main/java/org/monksanctum/xand11/comm/AuthManager.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.comm;

import android.text.TextUtils;
import android.util.Log;

import org.monksanctum.xand11.XServerInfo;
import org.monksanctum.xand11.XService;
import org.monksanctum.xand11.hosts.HostsManager;

import java.net.Socket;

/**
 * Handles the auth for the server.
 */
public class AuthManager {

    private static final String TAG = "AuthManager";
    private static final boolean DEBUG = XService.COMM_DEBUG;

    private static final boolean CARE_ABOUT_AUTH = false;

    public static final byte CONN_FAILED = 0;
    public static final byte CONN_SUCCEEDED = 1;
    public static final byte CONN_AUTH = 2;

    private final XServerInfo mInfo;
    // TODO: Actually use hosts.
    private final HostsManager mHosts;

    public AuthManager(XServerInfo info, HostsManager hostsManager) {
        mInfo = info;
        mHosts = hostsManager;
    }

    public boolean handleAuth(XProtoReader reader, XProtoWriter writer, Socket connection)
            throws XProtoReader.ReadException, XProtoWriter.WriteException {
        int majorVersion = reader.readCard16();
        int minorVersion = reader.readCard16();
        int n = reader.readCard16();
        int d = reader.readCard16();
        reader.readPadding(2);
        String authProtoName = reader.readPaddedString(n);
        String authProtoData = reader.readPaddedString(d);
        String host = connection.getInetAddress().getHostAddress();

        if (DEBUG) Log.d(TAG, "Incoming connection, host: " + host
                + " protocol: " + majorVersion + "," + minorVersion
                + " auth: " + authProtoName + " auth data: " + authProtoData);

        if (!CARE_ABOUT_AUTH || TextUtils.isEmpty(authProtoName)) {
            acceptConnection(writer);
            return true;
        } else {
            rejectConnection(writer, "Unknown auth " + authProtoName);
            return false;
        }
    }

    private void acceptConnection(XProtoWriter writer) throws XProtoWriter.WriteException {
        if (DEBUG) Log.d(TAG, "Writing accept");
        writer.writeByte(CONN_SUCCEEDED);
        writer.writePadding(1);
        writer.writeCard16(XService.MAJOR_VERSION);
        writer.writeCard16(XService.MINOR_VERSION);
        PacketWriter packet = new PacketWriter(writer);
        mInfo.write(packet);
        if (DEBUG) Log.d(TAG, "Packet length: " + packet.get4Length());
        writer.writeCard16(packet.get4Length());
        packet.flush();
    }

    private void rejectConnection(XProtoWriter writer, String reason)
            throws XProtoWriter.WriteException {
        if (DEBUG) Log.d(TAG, "Writing reject");
        writer.writeByte(CONN_FAILED);
        writer.writeByte((byte) reason.length());
        writer.writeCard16(XService.MAJOR_VERSION);
        writer.writeCard16(XService.MINOR_VERSION);
        writer.writeCard16((reason.length() + 3) / 4);
        writer.writePaddedString(reason);
    }
}

File: core/src/main/java/com/monksanctum/xand11/core/comm/AuthManager.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.comm

import org.monksanctum.xand11.core.*
import org.monksanctum.xand11.core.XServerInfo
import org.monksanctum.xand11.hosts.HostsManager

/**
 * Handles the auth for the server.
 */
class AuthManager(private val mInfo: XServerInfo, // TODO: Actually use hosts.
                  private val mHosts: HostsManager) {

    @Throws(XProtoReader.ReadException::class, XProtoWriter.WriteException::class)
    fun handleAuth(reader: XProtoReader, writer: XProtoWriter, connection: Socket): Boolean {
        val majorVersion = reader.readCard16()
        val minorVersion = reader.readCard16()
        val n = reader.readCard16()
        val d = reader.readCard16()
        reader.readPadding(2)
        val authProtoName = reader.readPaddedString(n)
        val authProtoData = reader.readPaddedString(d)
        val host = connection.getHostAddress()
        //inetAddress.hostAddress

        if (DEBUG) {
            Platform.logd(TAG, "Incoming connection, host: " + host
                    + " protocol: " + majorVersion + "," + minorVersion
                    + " auth: " + authProtoName + " auth data: " + authProtoData)
        }

        if (!CARE_ABOUT_AUTH || Platform.isTextEmpty(authProtoName)) {
            acceptConnection(writer)
            return true
        } else {
            rejectConnection(writer, "Unknown auth $authProtoName")
            return false
        }
    }

    @Throws(XProtoWriter.WriteException::class)
    private fun acceptConnection(writer: XProtoWriter) {
        if (DEBUG) Platform.logd(TAG, "Writing accept")
        writer.writeByte(CONN_SUCCEEDED)
        writer.writePadding(1)
        writer.writeCard16(MAJOR_VERSION)
        writer.writeCard16(MINOR_VERSION)
        val packet = PacketWriter(writer)
        mInfo.write(packet)
        if (DEBUG) Platform.logd(TAG, "Packet length: ${packet.WordCount}")
        writer.writeCard16(packet.WordCount)
        packet.flush()
    }

    @Throws(XProtoWriter.WriteException::class)
    private fun rejectConnection(writer: XProtoWriter, reason: String) {
        if (DEBUG) Platform.logd(TAG, "Writing reject")
        writer.writeByte(CONN_FAILED)
        writer.writeByte(reason.length.toByte())
        writer.writeCard16(MAJOR_VERSION)
        writer.writeCard16(MINOR_VERSION)
        writer.writeCard16((reason.length + 3) / 4)
        writer.writePaddedString(reason)
    }

    companion object {

        private val TAG = "AuthManager"
        private val DEBUG = COMM_DEBUG

        private val CARE_ABOUT_AUTH = false

        val CONN_FAILED: Byte = 0
        val CONN_SUCCEEDED: Byte = 1
        val CONN_AUTH: Byte = 2
    }
}
